src/ddsrt/include/dds/ddsrt/hopscotch.h

```diff
@@ -32,42 +47,172 @@ extern "C" {
+ * 
+ * @param[in] rt the hash table
+ * @param[in] keyobject is the object with which to do the lookup
+ * @return pointer to the matching element, NULL if failed
+ */
 DDS_EXPORT void *ddsrt_hh_lookup (const struct ddsrt_hh * __restrict rt, const void * __restrict keyobject) ddsrt_nonnull_all;
+
+/**
+ * @brief Add an element to the hash table.
+ * 
+ * It is not allowed to add duplicates to the hash table. Trying to do so results in failure.
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] data user data to add
+ * @return int meant to be interpreted as bool (1 is success, 0 is failure)
+ * @see @ref ddsrt_hh_remove
+ */
 DDS_EXPORT int ddsrt_hh_add (struct ddsrt_hh * __restrict rt, void * __restrict data) ddsrt_nonnull_all;
+
+/**
+ * @brief Remove an element from the hash table.
+ * 
+ * It only uses the key of the passed object and as such it will also work if you pass a dummy object.
+ * However, you still need to free the memory of the object after removing it from the hash table.
+ * Since this means you need to have the pointer to the actual object anyway, you might as well use that.
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] keyobject user data to remove
+ * @return int meant to be interpreted as bool (1 is success, 0 is failure)
+ * @see @ref ddsrt_hh_add
+ */
 DDS_EXPORT int ddsrt_hh_remove (struct ddsrt_hh * __restrict rt, const void * __restrict keyobject) ddsrt_nonnull_all;
+
+/**
+ * @brief Like @ref ddsrt_hh_add, but without returning success/failure result.
+ * 
+ * Only use if you're certain that the element doesn't already exist.
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] data user data to add
+ * @see @ref ddsrt_hh_remove
+ */
 DDS_EXPORT void ddsrt_hh_add_absent (struct ddsrt_hh * __restrict rt, void * __restrict data) ddsrt_nonnull_all;
+
+/**
+ * @brief Like @ref ddsrt_hh_remove, but without returning success/failure result.
+ * 
+ * Only use if you're certain that the element exists.
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] keyobject user data to remove
+ * @see @ref ddsrt_hh_add_absent
+ */
 DDS_EXPORT void ddsrt_hh_remove_present (struct ddsrt_hh * __restrict rt, void * __restrict keyobject) ddsrt_nonnull_all;
+
+/**
+ * @brief Walk the hash table and apply a user defined function to each node.
+ * 
+ * The walk function 'f' receives in parameter 'a' the pointer to user data, and has 'f_arg' as an optional extra argument.
+ * It is allowed to modify the user data. The @ref ddsrt_hh_enum is useful to free user data prior to calling @ref ddsrt_hh_free.
+ * 
+ * @param[in,out] rt the hash table
+ * @param[in] f user defined walk function to apply to each element
+ * @param[in] f_arg extra argument for walk function
+ * @see @ref ddsrt_hh_iter_next
+ */
 DDS_EXPORT void ddsrt_hh_enum (struct ddsrt_hh * __restrict rt, void (*f) (void *a, void *f_arg), void *f_arg) ddsrt_nonnull ((1, 2)); /* may delete a */
```

Comment:
The "dangling" comment "may delete a" should be moved into the doxygen comment, I'd say. It should also be made clear that it is `f` that is allowed to delete a. `enum` by itself doesn't modify anything.